docs/concepts/pull-dependencies.md

```diff
@@ -0,0 +1,202 @@
+                "dep_b", "dep_c"
+            ]
+        },
+        "dep_b": {
+            "type": "pull",
+            "id": "yhgn4-myaaa-aaaaa-aabta-cai"
+        },
+        "dep_c": {
+            "type": "pull",
+            "id": "yahli-baaaa-aaaaa-aabtq-cai"
+        }
+    }
+}
+```
+
+### 2. Pull the dependencies using `dfx deps pull`
+
+Running `dfx deps pull --network ic` will:
+
+1. resolve the dependency graph by fetching `dfx:deps` metadata recuresively;
+2. download wasm of all direct and indirect dependencies from `dfx:wasm_url` into shared cache;
+3. hash check the downloaded wasm against `dfx:wasm_hash` metadata or the hash of the mainnet running canister;
+4. extract `candid:args`, `candid:service`, `dfx:init` from the downloaded wasm;
+5. create `deps/` folder in project root;
+6. save `candid:service` of direct dependencies as `deps/<NAME>.did`;
+7. save `deps/pulled.json` which contains major info of all direct and indirect dependencies;
+
+For the example project, you will find following files in `deps/`:
+
+- `dep_b.did` and `dep_c.did`: candid files that can be imported by "app";
+- `pulled.json` which has follwing content:
+
+```
+{
+  "canisters": {
+    "yofga-2qaaa-aaaaa-aabsq-cai": {
+      "name": null,
+      "deps": [],
+      "wasm_url": "example.com/a.wasm",
+      "wasm_hash": "e9b8ba2ad28fa1403cf6e776db531cdd6009a8e5cac2b1097d09bfc65163d56f",
+      "dfx_init": "Nat",
+      "candid_args": "(nat)"
+    },
+    "yhgn4-myaaa-aaaaa-aabta-cai": {
+      "name": "dep_b",
+      "deps": [
+        "yofga-2qaaa-aaaaa-aabsq-cai"
+      ],
+      "wasm_url": "example.com/b.wasm",
+      "wasm_hash": "f607c30727b0ee81317fc4547a8da3cda9bb9621f5d0740806ef973af5b479a2",
+      "dfx_init": null,
+      "candid_args": "()"
+    },
+    "yahli-baaaa-aaaaa-aabtq-cai": {
+      "name": "dep_c",
+      "deps": [
+        "yofga-2qaaa-aaaaa-aabsq-cai"
+      ],
+      "wasm_url": "example.com/c.wasm",
+      "wasm_hash": "016df9800dc5760785646373bcb6e6bb530fc17f844600991a098ef4d486cf0b",
+      "dfx_init": "Nat",
+      "candid_args": "(nat)"
+    }
+  }
+}
+```
+
+There are three dependencies:
+
+- "yhgn4-myaaa-aaaaa-aabta-cai": "dep_b" in `dfx.json`;
+- "yahli-baaaa-aaaaa-aabtq-cai": "dep_c" in `dfx.json`;
+- "yofga-2qaaa-aaaaa-aabsq-cai": an indirect dependency that both "dep_b" and "dep_c" depend on;
+
+### 3. Set init arguments using `dfx deps init`
+
+Running `dfx deps init` will iterate all dependencies in `pulled.json`, set empty argument for the ones need no init argument and print the list of dependencies that require init argument.
+
+Then running `dfx deps init <CANISTER> --argument <ARGUMENT>` to set init argument for individual dependency.
+
+The init arguments will be recorded in `deps/init.json`.
+
+For our example, we should run:
+
+```
+> dfx deps init
+WARN: The following canister(s) require an init argument. Please run `dfx deps init <NAME/PRINCIPAL>` to set them individually:
+yofga-2qaaa-aaaaa-aabsq-cai
```

Comment:
Note that the `dfx:init` metadata in the canister that referenced this indirect dependency does have a name for this dependency, but we discarded it.  Can display it here in an unambiguous way?

Reply:
In my e2e example project, both `b` and `c` depends on `a`. It's possible that `b` and `c` set different names for `a` in their `dfx:deps`. Which name should we use here?

Another idea is to remove the `name` from `dfx:deps` so that it only contains a list of principals.

Reply:
I wondered what to do about the potentially different names too, but I don't have a clear answer either.